aws/resource_aws_kinesis_analytics_application.go

```diff
@@ -0,0 +1,1521 @@
+		ApplicationName: aws.String(name),
+	}
+
+	if v, ok := d.GetOk("code"); ok && v.(string) != "" {
+		createOpts.ApplicationCode = aws.String(v.(string))
+	}
+
+	if v, ok := d.GetOk("cloudwatch_logging_options"); ok {
+		clo := v.([]interface{})[0].(map[string]interface{})
+		cloudwatchLoggingOption := createCloudwatchLoggingOption(clo)
+		createOpts.CloudWatchLoggingOptions = []*kinesisanalytics.CloudWatchLoggingOption{cloudwatchLoggingOption}
+	}
+
+	if v, ok := d.GetOk("inputs"); ok {
+		i := v.([]interface{})[0].(map[string]interface{})
+		inputs := createInputs(i)
+		createOpts.Inputs = []*kinesisanalytics.Input{inputs}
+	}
+
+	if v, ok := d.GetOk("outputs"); ok {
+		o := v.([]interface{})[0].(map[string]interface{})
+		outputs := createOutputs(o)
+		createOpts.Outputs = []*kinesisanalytics.Output{outputs}
+	}
+
+	err := resource.Retry(1*time.Minute, func() *resource.RetryError {
+		_, err := conn.CreateApplication(createOpts)
+		if err != nil {
+			if isAWSErr(err, kinesisanalytics.ErrCodeInvalidArgumentException, "Kinesis Analytics service doesn't have sufficient privileges") {
+				return resource.RetryableError(err)
+			}
+			return resource.NonRetryableError(err)
+		}
+		return nil
+	})
+	if err != nil {
+		return fmt.Errorf("Unable to create Kinesis Analytics application: %s", err)
+	}
+
+	return resourceAwsKinesisAnalyticsApplicationUpdate(d, meta)
+}
+
+func resourceAwsKinesisAnalyticsApplicationRead(d *schema.ResourceData, meta interface{}) error {
+	conn := meta.(*AWSClient).kinesisanalyticsconn
+	name := d.Get("name").(string)
+
+	describeOpts := &kinesisanalytics.DescribeApplicationInput{
+		ApplicationName: aws.String(name),
+	}
+	resp, err := conn.DescribeApplication(describeOpts)
+	if err != nil {
+		if awsErr, ok := err.(awserr.Error); ok {
+			if awsErr.Code() == "ResourceNotFoundException" {
+				d.SetId("")
+				return nil
+			}
+			return fmt.Errorf("[WARN] Error reading Kinesis Analytics Application: \"%s\", code: \"%s\"", awsErr.Message(), awsErr.Code())
+		}
+		return err
+	}
+
+	d.SetId(aws.StringValue(resp.ApplicationDetail.ApplicationARN))
+	d.Set("name", aws.StringValue(resp.ApplicationDetail.ApplicationName))
+	d.Set("arn", aws.StringValue(resp.ApplicationDetail.ApplicationARN))
+	d.Set("code", aws.StringValue(resp.ApplicationDetail.ApplicationCode))
+	d.Set("create_timestamp", aws.TimeValue(resp.ApplicationDetail.CreateTimestamp).Format(time.RFC3339))
+	d.Set("description", aws.StringValue(resp.ApplicationDetail.ApplicationDescription))
+	d.Set("last_update_timestamp", aws.TimeValue(resp.ApplicationDetail.LastUpdateTimestamp).Format(time.RFC3339))
+	d.Set("status", aws.StringValue(resp.ApplicationDetail.ApplicationStatus))
+	d.Set("version", int(aws.Int64Value(resp.ApplicationDetail.ApplicationVersionId)))
+
+	if err := d.Set("cloudwatch_logging_options", getCloudwatchLoggingOptions(resp.ApplicationDetail.CloudWatchLoggingOptionDescriptions)); err != nil {
```

Comment:
Since this project uses a very flat Go package structure, we prefer to include the AWS service name in the function name. We also generally call these functions "flatten" functions. e.g. `flattenKinesisAnalyticsCloudwatchLoggingOptions`

Same for the below `getX` function names.